subt/odoimuloc.py

```diff
@@ -0,0 +1,84 @@
+import collections
+import logging
+import math
+import threading
+
+import osgar.lib.quaternion as quaternion
+
+g_logger = logging.getLogger(__name__)
+
+Pose2d = collections.namedtuple("Pose2d", ("x", "y", "heading"))
+
+class Localization:
+    def __init__(self, config, bus):
+        self.bus = bus
+        # outputs
+        bus.register('pose3d')
+        # inputs: origin, orientation, odom
+        self.xyz = None
+        self.orientation = None
+        self.last_odom = None
+        self.origin_xyz = None
+        self.origin_orientation = None
+        self.origin_error = False
+
+    def start(self):
+        self.thread = threading.Thread(target=self.main)
+        self.thread.start()
+
+    def join(self, timeout=None):
+        self.thread.join(timeout)
+
+    def on_origin(self, data):
+        if len(data) == 8:
+            self.xyz = data[1:4]
+            self.origin_xyz = data[1:4]
+            self.origin_orientation = data[4:]
+        else:
+            self.origin_error = True
+
+    def on_orientation(self, dt, orientation):
+        self.orientation = orientation
+
+    def on_odom(self, dt, x, y, heading):
+        assert self.xyz is not None
+        if self.orientation is None:
+            return
+
+        odom = Pose2d(x / 1000.0, y / 1000.0, math.radians(heading / 100.0))
+        if self.last_odom is not None:
+            dist = math.hypot(odom.x - self.last_odom.x, odom.y - self.last_odom.y)
+            direction = ((odom.x - self.last_odom.x) * math.cos(self.last_odom.heading) +
+                         (odom.y - self.last_odom.y) * math.sin(self.last_odom.heading))
+            if direction < 0:
+                dist = -dist
+        else:
+            dist = 0.0
+
+        self.last_odom = odom
+        dist3d = quaternion.rotate_vector([dist, 0, 0], self.orientation)
+        self.xyz = [a + b for a, b in zip(self.xyz, dist3d)]
+        self.bus.publish('pose3d', [self.xyz, self.orientation])
+
+    def main(self):
+        # wait for valid origin
+        while True:
+            dt, channel, data = self.bus.listen()
+            if channel == "origin":
```

Comment:
this functionality could be already inherited?

Reply:
well, calling `super().update()` or `self.bus.listen()` is just as long :man_shrugging: